src/renderers/webgl/WebGLBindingStates.js

```diff
@@ -0,0 +1,393 @@
+﻿/**
+ * @author Mugen87 / https://github.com/Mugen87
+ * @author Takahiro / https://github.com/takahirox
+ */
+
+function WebGLBindingStates( gl, extensions, state, attributes, capabilities ) {
+
+	var maxVertexAttributes = gl.getParameter( gl.MAX_VERTEX_ATTRIBS );
+
+	var extension = extensions.get( 'OES_vertex_array_object' );
+	var vaoAvailable = capabilities.isWebGL2 || extension !== null;
+
+	var bindingStates = new Map();
+
+	var defaultState = createBindingState( null );
+	var currentState = defaultState;
+
+	function setup( material, program, geometry, hasMorphTarget ) {
+
+		var state = getBindingState( geometry, program, material );
+
+		if ( currentState !== state ) {
+
+			currentState = state;
+			bindVertexArrayObject( currentState.object );
+
+		}
+
+		var updateBuffers = hasMorphTarget;
```

Comment:
Update vertex attribute if morph targets are present so far because current morph targets rendering path requires attributes update. We could find solution in the future.

Reply:
I realized I can remove `hasMorphTarget` because geometry version increments in `WebGLMorphTargets` when morph attributes are added/removed.

And I optimized morph attributes addition/removal in `WebGLMorphTargets`. They're added/removed only when necessary. So I accomplished to narrow the case where VAO needs to be updated even with morph targets.